src/coreComponents/physicsSolvers/fluidFlow/GeochemicalModel.cpp

```diff
@@ -0,0 +1,819 @@
+          localMatrix.addToRow< serialAtomic >( localAccumDOF[i],
+                                                localAccumDOF.data(),
+                                                localAccumJacobian[i],
+                                                m_numDofPerCell );
+        }
+
+      }
+
+    } );
+  } );
+
+}
+
+void GeochemicalModel::AssembleFluxTerms( real64 const time_n,
+                                          real64 const dt,
+                                          DomainPartition const & domain,
+                                          DofManager const & dofManager,
+                                          CRSMatrixView< real64, globalIndex const > const & localMatrix,
+                                          arrayView1d< real64 > const & localRhs )
+{
+  GEOSX_UNUSED_VAR( time_n );
+  GEOSX_UNUSED_VAR( dt );
+  GEOSX_UNUSED_VAR( domain );
+  GEOSX_UNUSED_VAR( dofManager );
+  GEOSX_UNUSED_VAR( localMatrix );
+  GEOSX_UNUSED_VAR( localRhs );
+}
+
+void GeochemicalModel::ApplyBoundaryConditions(real64 const time,
+                                               real64 const dt,
+                                               DomainPartition & domain,
+                                               DofManager const & dofManager,
+                                               CRSMatrixView< real64, globalIndex const > const & localMatrix,
+                                               arrayView1d< real64 > const & localRhs )
+{
+  GEOSX_UNUSED_VAR( time );
+  GEOSX_UNUSED_VAR( dt );
+  GEOSX_UNUSED_VAR( domain );
+  GEOSX_UNUSED_VAR( dofManager );
+  GEOSX_UNUSED_VAR( localMatrix );
+  GEOSX_UNUSED_VAR( localRhs );
+
+}
+
+
+real64
+GeochemicalModel::
+CalculateResidualNorm( DomainPartition const & domain,
+                       DofManager const & dofManager,
+                       arrayView1d< real64 const > const & localRhs )
+{
+
+  MeshLevel const & mesh = *(domain.getMeshBody(0)->getMeshLevel(0));
+
+  string const dofKey = dofManager.getKey( viewKeyStruct::pressureString );  
+  globalIndex const rankOffset = dofManager.rankOffset();
+
+  // compute the norm of local residual scaled by cell pore volume
+  real64 localResidualNorm = 0.0;
+
+  forTargetSubRegionsComplete( mesh,
+                               [&]
+                               ( localIndex const,
+                                 localIndex const er,
+                                 localIndex const esr,
+                                 ElementRegionBase const & GEOSX_UNUSED_PARAM( region ),
+                                 ElementSubRegionBase const & subRegion )
+  {
+
+    arrayView1d<globalIndex const> const & dofNumber = subRegion.getReference< array1d<globalIndex> >( dofKey );
+    arrayView1d<integer const> const & elemGhostRank = m_elemGhostRank[er][esr];
+
+    localIndex const subRegionSize = subRegion.size();
+    for ( localIndex a = 0; a < subRegionSize; ++a )
+    {
+
+      if (elemGhostRank[a] < 0)
```

Comment:
Is this check required any longer? Or....should we be looping over the residual vector instead?

Reply:
Just looping over the residual vector is fine as long as the solver is iteratively coupled and all you need as an unscaled L2 norm. In fact, we could add that as a base implementation in `SolverBase::CalculateResidualNorm()`.

If the solver is just a block in a fully implicit multiphysics solve, then we have to do a mesh-based loop and the ghost check is required... or you could check `lid` below to be in range, but I don't like it as it can hide bugs.